### ISSUE TITLE
Variables can be accessed by any route

### ISSUE THREAD

**Issue by traviops** [mockoon/mockoon#920]
Would be super useful if it was possible to access variables regardless of routes/responses that they were set to initially.

Currently it's only possible to set a variable and get its value in the same response which it was set initially.

**Comment by traviops** [comment]
Up.

**Comment by 255kb** [comment]
This should be the role of the data buckets which are shared. 
I agree that you can only read them right now. There is an [opened issue](https://github.com/mockoon/mockoon/issues/1098) to allow arbitrary modification to data buckets. It will solve your issue. 
I will close this issue as already partly covered and duplicate.

**Comment by traviops** [comment]
@255kb Actually I think global variables have a different usage than data buckets. I say that looking at the scenario where I need to store data from requests to use in other requests, but I don't need to maintain them, only while the mock server is running.

Using data buckets as global variables can be very messy because in data buckets I usually create static values, if I were to store dynamic data from requests inside data buckets it would make everything sort of cluttered.

With global variables approach this issue wouldn't happen.

Don't you agree?

**Comment by 255kb** [comment]
After thinking about it for a while I think I agree with you.
The data bucket system is here, convenient, but generating content once and for all at server start (it can be still be manipulated later using CRUD routes). It's true that using it as a variable system to store, let's say, "route no. 1 URL param during last request", could be a bit confusing as it is not anymore the DB-like storage system we first intended. 

Maybe a separated tab could be created with a key/value system where we can define specific pieces of data to fetch every time. 

Let's keep this issue open and see if there is more interest or feedback.

**Comment by LukMas** [comment]
> Let's keep this issue open and see if there is more interest or feedback.
I would like to have this feature too. As I wrote on #1098, this feature would allow more complex interactions that actually are not possible with the data bucket.

**Comment by realsaid84** [comment]
I was a keen supporter of arbitrary changes to data buckets but upon reflection a key/value system that is global in nature as @255kb mentioned is very useful in setting up a scenario with multiple endpoints, for example:


> A Transaction lifecycle scenario

1. Create transaction  POST `/transactions` 

```json
{
   "quantity": 1,
  "price_id": "pri_01gsz98e27ak2tyhexptwc58yk",
  "customer_id": "ctm_01h8441jn5pcwrfhwh78jqt8hk",
  "address_id": "add_01h848pep46enq8y372x7maj0p",
  "currency_code": "USD",
  "collection_mode": "manual",
   "purchase_order_number": "PO-123",
  }
```
Response 
```json
{
"transactionId": "959c78b0-5e89-42fc-b121-6cb3073ea1f9"
}
```

_The response transactionId can be used in setting a global key/value entry against the entire request body in the response template_

eg. 
```
{
{{{setGlobalVar globalVariableName (guid) (bodyRaw)}}}
"transaction": {{{@globalVariableName.key}}} 
}
```


2. GET  **transaction** `/transactions/959c78b0-5e89-42fc-b121-6cb3073ea1f9
```
{
"transactionId": {{{@globalVariableName.key}}} ,
{{{@globalVariableName.value}}} 
}
```

_Which will yield the following 200 OK response:_
```json
{
  "transactionId": "959c78b0-5e89-42fc-b121-6cb3073ea1f9",
   "quantity": 1,
  "price_id": "pri_01gsz98e27ak2tyhexptwc58yk",
  "customer_id": "ctm_01h8441jn5pcwrfhwh78jqt8hk",
  "address_id": "add_01h848pep46enq8y372x7maj0p",
  "currency_code": "USD",
  "collection_mode": "manual",
   "purchase_order_number": "PO-123",
  }
```

3. DELETE  **transaction** `/transactions/959c78b0-5e89-42fc-b121-6cb3073ea1f9
```
{{{unsetGlobalVar globalVariableName}}}
```

This clears the global variable or can be unset based on some invalidation conditions or a specific timeout setting.

**Comment by 255kb** [comment]
@realsaid84 This is the kind of use case I have in mind too. 
I first imagined a separated interface (like a new tab), where you could set the variables by selecting what to assign from which route, like a body property, etc.
But this could be a bit of a hassle and or require a DSL. 
I find adding a new helper probably a better solution. It's easier to code, and already contextualized. 

I will add this on the roadmap.

**Comment by LukMas** [comment]
@255kb I don't know if it's planned already like I need/would see, but would it be possible to make every change independent from the verb of the route? I mean: 
- inside a GET `/api/currentUser?username=myUsername` I create/store `{setGlobalVar g_username myUsername}`
- I get the current username from the GET `/api/user/username`
- Within the GET `/api/switchUsername?username=otherUsername` I create `{setGlobalVar g_old_username myUsername}` and update `{setGlobalVar g_username otherUsername}`
- Within the GET `/api/resetUsername` I restore the original username into `g_username` and destroy `g_old_username`

This would allow to change a global variable from any route/verb combination.

**Comment by 255kb** [comment]
@LukMas I don't see why we would limit the use of helpers to any verb. The idea of Mockoon is to be permissive and we already can add any header, body, etc. to any verb/route. So, yes it would work as you describe I guess 🙂

**Comment by traviops** [comment]
> @realsaid84 This is the kind of use case I have in mind too. I first imagined a separated interface (like a new tab), where you could set the variables by selecting what to assign from which route, like a body property, etc. But this could be a bit of a hassle and or require a DSL. I find adding a new helper probably a better solution. It's easier to code, and already contextualized.
> 
> I will add this on the roadmap.

Thank you! This will be a great feature allowing a lot of use cases!

**Comment by 255kb** [comment]
I did a very simple implementation like suggested, with two new helpers `setGlobalVar` and `getGlobalVar`. It will be in December release.

**Comment by realsaid84** [comment]
@255kb , thanks a lot for this implementation, looking forward to the December release.

**Comment by LukMas** [comment]
@255kb super!
Thank you very much.

**Comment by 255kb** [comment]
@realsaid84 @LukMas You are welcome, I hope it brings you joy :D
Then we can iterate if needed.

**Comment by 255kb** [comment]
📦 This is now released in [v6.1.0](https://mockoon.com/releases/6.1.0/).
